File: src/game/inner/timer.rs
pub struct Timer {
    timespan: f64,
    current: f64,
}

impl Timer {
    pub fn new(timespan: f64) -> Self {
        Self {
            timespan,
            current: timespan,
        }
    }

    pub fn tick(&mut self, dt: f64) -> bool {
        self.current -= dt;
        if self.current < 0.0 {
            self.current = self.timespan;
            true
        } else {
            false
        }
    }
}

File: haskell/app/Game/Timer.hs
module Game.Timer (Timer, mkTimer, tick) where

data Timer = Timer
  { timespan :: Double
  , current  :: Double
  } deriving Show

mkTimer :: Double -> Timer
mkTimer span = Timer
  { timespan = span
  , current  = span
  }

tick :: Double -> Timer -> (Timer, Bool)
tick dt t =
  let newCurrent = current t - dt
  in
    if newCurrent < 0.0
    then (t { current = timespan t }, True )
    else (t { current = newCurrent }, False)